core/src/main/java/com/google/errorprone/bugpatterns/nullness/RedundantNullCheck.java

```diff
@@ -0,0 +1,109 @@
+import com.google.errorprone.bugpatterns.nullness.NullnessUtils.NullCheck;
+import com.google.errorprone.dataflow.nullnesspropagation.Nullness;
+import com.google.errorprone.dataflow.nullnesspropagation.NullnessAnnotations;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.util.ASTHelpers;
+
+import com.sun.source.tree.BinaryTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+
+import java.util.Optional;
+
+import javax.lang.model.element.ElementKind;
+
+@BugPattern(
+    summary = "Explicit null check on a variable or method call that is not @Nullable within a @NullMarked scope.",
+    severity = WARNING)
+public class RedundantNullCheck extends BugChecker implements BinaryTreeMatcher {
+
+  @Override
+  public Description matchBinary(BinaryTree tree, VisitorState state) {
+    NullCheck nullCheck = NullnessUtils.getNullCheck(tree);
+    if (nullCheck == null) {
+      return NO_MATCH;
+    }
+
+    VarSymbol varSymbol = nullCheck.varSymbolButUsuallyPreferBareIdentifier();
+    if (varSymbol != null) {
+      if (!NullnessUtils.isInNullMarkedScope(varSymbol, state)) {
+        return NO_MATCH;
+      }
+      if (NullnessUtils.isAlreadyAnnotatedNullable(varSymbol)) {
+        return NO_MATCH;
+      }
+
+      VariableTree varDecl = NullnessUtils.findDeclaration(state, varSymbol);
+
+      if ((varSymbol.getKind() == ElementKind.LOCAL_VARIABLE
+          || varSymbol.getKind() == ElementKind.RESOURCE_VARIABLE) && varDecl != null) {
+
+        if (varDecl.getInitializer() == null) {
+          return NO_MATCH;
+        }
+
+        Tree initializer = varDecl.getInitializer();
+
+        if (initializer.getKind() == Tree.Kind.METHOD_INVOCATION) {
+          MethodInvocationTree methodInvocation = (MethodInvocationTree) initializer;
+          MethodSymbol methodSymbol = ASTHelpers.getSymbol(methodInvocation);
+          if (methodSymbol != null && isEffectivelyNullable(methodSymbol, state)) {
+            return NO_MATCH;
+          }
+        } else {
+          return NO_MATCH;
+        }
+      }
+    } else {
+      MethodSymbol methodSymbol = nullCheck.methodSymbol();
+      if (methodSymbol == null) {
+        return NO_MATCH;
+      }
+      if (isEffectivelyNullable(methodSymbol, state)) {
+        return NO_MATCH;
+      }
+    }
+    return buildDescription(tree).build();
+  }
+
+  private static boolean isEffectivelyNullable(MethodSymbol methodSymbol, VisitorState state) {
+    Optional<Nullness> returnTypeNullness = NullnessAnnotations.fromAnnotationsOn(methodSymbol);
+    if (returnTypeNullness.isPresent()) {
+      // Explicit @Nullable or @NonNull on the return type
+      return returnTypeNullness.get() == Nullness.NULLABLE;
+    }
+    // No explicit annotation on return type.
```

Comment:
Similarly to the other type-variable case above:

```suggestion
    if (methodSymbol.getReturnType().getKind() == TYPEVAR) {
      return true;
    }
    // No explicit annotation on return type.
```